### ISSUE TITLE
Rendering artifacts in ListView

### ISSUE THREAD

**Issue by CodeDevAM** [AvaloniaUI/Avalonia#13607]
## Describe the bug
I am experiencing strange rendering artefacts when using a `ListView`.

## To Reproduce

See attached minimal example.
[ListViewAutoScrollBug.zip](https://github.com/AvaloniaUI/Avalonia/files/13349670/ListViewAutoScrollBug.zip)

## Screenshots

https://github.com/AvaloniaUI/Avalonia/assets/125025268/56b2cfc4-e568-4f26-bc02-63c9ad0ed28f

## Environment 

- OS: Windows 11
- Avalonia-Version: 11.0.5

## Additional context
I want to render a hierarchical tree in a ListView. Every change of the tree triggers a redraw of the ListView so that the content of the ListView is recreated every time a node is expanded or collapsed.

For illustration purposes in this demo every expand of a node creates 20 new children while the last of them shall be selected and become visible. (`AutoScrollToSelectedItem`) This works most of the time. But sometimes the ListView "overdraws" the item with some old content that does not disappear anymore. In these cases the autoscroll does not work either.

**Comment by timunie** [comment]
consider using TreeDataGrid

**Comment by rabbitism** [comment]
I can also confirm TreeDataGrid has virtualization issue.

**Comment by timunie** [comment]
y, the underlaying issue is applying to all virtualized controls. So if anyone finds the root cause, a PR is more than welcome. Maybe @grokys has some ideas but he is more than busy rn.

**Comment by CodeDevAM** [comment]
I don't know if this is helpful but I noticed that it is reproducable that exacly the same node seems to be affected every time you run the app. (Just expand the last node until the issue appears)

And the it seems like a "broken" node is added additionally to the other nodes before the actual new nodes are added. This broken node also appears in the logical tree.

![grafik](https://github.com/AvaloniaUI/Avalonia/assets/125025268/834618d4-d2f2-4ded-93d1-6d2fccfde2bd)

**Comment by CodeDevAM** [comment]
The issue seems to occur only when `AutoScrollToSelectedItem` is true.

**Comment by CodeDevAM** [comment]
At least in my example it only occurs when the last item gets selected by code.

**Comment by CodeDevAM** [comment]
In my tests the issue disappeared when `UseLayoutRounding` is set to `false` on `ListBox`.

**Comment by grokys** [comment]
Should be fixed by #13765 if you'd like to test.

**Comment by CodeDevAM** [comment]
I have tested the attached example with 11.0.6 and the issue is still present.

**Comment by timunie** [comment]
@CodeDevAM can you also check nightly? It has fixes that 11.0.6 doesn't have

**Comment by CodeDevAM** [comment]
No difference with 11.1.999-cibuild0043367-beta

**Comment by timunie** [comment]
Reopening for now. 

/cc @grokys

**Comment by CodeDevAM** [comment]
The issue is still present with 11.0.9.

Unfortunately this is a big blocker for my app.

Maybe this helps finding the root cause: I have provided an _Reset View_ button the the user for the situation when the issue occurs in my app. So the user can manually recreate the ListView Control. Then issue does not appears for the newly created control. But it may appear again after some modifications to the list.

**Comment by CodeDevAM** [comment]
My current investigation results:

- Occurance depands on the size of Controls. A slightly different size results in different bahaviour.

https://github.com/AvaloniaUI/Avalonia/assets/125025268/b6cc4178-ad80-49ff-b7a0-1a5b3120ab02


https://github.com/AvaloniaUI/Avalonia/assets/125025268/7e4e594d-66fe-470d-9277-da793c012b8e

- `ScrollIntoView()` from `VirtualizingStackPanel` seems to add Items that are not initialized correctly. Not all `VisualChildren` are cleared corretly afterwards. This results in overlaying Controls.

- Different settings for `UseLayoutRounding` have an impact on the behaviour but the issue **does not** disappear.

**Comment by CodeDevAM** [comment]
I don't have a minimal reproduction scenario yet but `TreeDataGrid` seems to have the same or a very similar issue. (Version 11.0.2)
Since I don't have a solution for `AutoScrollToSelectedItem` with a `TreeDataGrid` yet the issue also occours independently of this feature.

**Comment by timunie** [comment]
@CodeDevAM I'd suggest you put your sample into `src/samples/Sandbox` and try to debug the virtulization logic in source directly, given you spend some time on that issue anyways. Maybe you find the right place to clean up recycle items.

**Comment by CodeDevAM** [comment]
That's exactly what I was doing. I spent many hours to figure out the root cause of this issue. (at least for a `ListBox`) Unfortunatly it's a very complicated piece of code and I was not able to figure it out so far.

The problem that I have is that changing the underlying collection of a `ListBox` causes a run through a hierarchy of Controls multiple times in different modes that changes the behaviour of the relevant functions slightly. In the end it is very hard to follow. Long story short: I think it needs an expert that is more familiar with the inner details of a `ListBox`.

**Comment by timunie** [comment]
@CodeDevAM if you give a kind of "epsilon" to the contains method, would that solve your issue? 

`Bounds.Contains(rect, 0.0001)` where Contains needs a new overload accepting that epsilon.

**Comment by CodeDevAM** [comment]
A check with a certain tolerance would be a good idea. 

Unfortunately it does not solve the issue. With my analysis I figured out so far that `_viewport` of the `VirtualizingStackPanel` is calculated differently in a good case and a bad case at
https://github.com/AvaloniaUI/Avalonia/blob/11f5203ab24515a4094d7140d0d757d5016e2007/src/Avalonia.Controls/VirtualizingStackPanel.cs#L422

As a result the `anchorIndex` is is not calculated correctly. https://github.com/AvaloniaUI/Avalonia/blob/11f5203ab24515a4094d7140d0d757d5016e2007/src/Avalonia.Controls/VirtualizingStackPanel.cs#L465

**Comment by timunie** [comment]
Another idea would be to use Inflate on the Bounds and friends 🤔

**Comment by MrJul** [comment]
I can reproduce the issue consistently when the item to select has a width larger than the `ListBox` (by resizing the window enough).

This causes the inner `ScrollViewer` (the `ListBox`'s one) to incorrectly handle the `RequestBringIntoView` for the item, preventing the outer one (inside `MainWindow)` from scrolling to the correct position. This in turn prevents the focused item from being part of the realized item list, duplicating it.

Here's a PR to fix that:

@CodeDevAM can you please confirm that this fixes your issue? Several of your videos don't seem to have the item larger than the viewport, so maybe there's still something else going on.

**Comment by maxkatz6** [comment]
That fix can be tested on 11.2.999-cibuild0046020-beta builds, just in case.

**Comment by MrJul** [comment]
Thank you for this easy repro, this really helped!

With the knowledge from #14900, I could immediately see that the cause was similar: the inner `ScrollViewer` is still incorrectly handling the event. This time it's because the new offset is nearing 0 but isn't zero (but still larger than the espilon used in `NearlyEquals`), due to rounding errors.

The offset is correctly coerced back to zero thought, but this wasn't taken into account inside `BringDescendantIntoView`.

#14913 should really fix it time :)

-----

Note that I believe that the `VirtualizingStackPanel` creating the focused element and bringing it into view should be completely reworked at some point. It's completely valid to cancel the `BringIntoView` for various reasons, effectively duplicating the focused item immediately.

**Comment by CodeDevAM** [comment]
Thank you very much for your effort! These are great news.

I can verify that build 11.2.999-cibuild0046045-beta (https://github.com/AvaloniaUI/Avalonia/pull/14913) fixes the issue with the minimal example app.

Please consider backporting https://github.com/AvaloniaUI/Avalonia/pull/14913 and https://github.com/AvaloniaUI/Avalonia/pull/14900 to `11.0.X`.
